grpc-server.go

```diff
@@ -788,137 +779,267 @@ func (multi *MultiEpoch) processSlotTransactions(
+						if txn.Slot < int(startSlot) || txn.Slot > int(endSlot) {
+							continue
+						}
+						slot := uint64(txn.Slot)
+						slotTxns[slot] = append(slotTxns[slot], txn)
+					}
 
-						txResp := new(old_faithful_grpc.TransactionResponse)
-						txResp.Transaction = new(old_faithful_grpc.Transaction)
-						{
-							pos, ok := txn.GetPositionIndex()
-							if ok {
-								txResp.Index = ptrToUint64(uint64(pos))
-								txResp.Transaction.Index = ptrToUint64(uint64(pos))
-							}
-							txResp.Transaction.Transaction, txResp.Transaction.Meta, err = getTransactionAndMetaFromNode(txn, epochHandler.GetDataFrameByCid)
+					// Process transactions slot by slot
+					for slot, txnsInSlot := range slotTxns {
+						for _, txn := range txnsInSlot {
+							tx, meta, err := parseTransactionAndMetaFromNode(txn, epochHandler.GetDataFrameByCid)
 							if err != nil {
-								return status.Errorf(codes.Internal, "Failed to get transaction: %v", err)
+								errChan <- status.Errorf(codes.Internal, "Failed to parse transaction from node: %v", err)
+								return
 							}
-							txResp.Slot = uint64(txn.Slot)
 
-							blocktimeIndex := epochHandler.GetBlocktimeIndex()
-							if blocktimeIndex != nil {
-								blocktime, err := blocktimeIndex.Get(uint64(txn.Slot))
-								if err != nil {
-									return status.Errorf(codes.Internal, "Failed to get blocktime: %v", err)
+							if !filterOutTxn(tx, meta) {
+								txResp := new(old_faithful_grpc.TransactionResponse)
+								txResp.Transaction = new(old_faithful_grpc.Transaction)
+								{
+									pos, ok := txn.GetPositionIndex()
+									if ok {
+										txResp.Index = ptrToUint64(uint64(pos))
+										txResp.Transaction.Index = ptrToUint64(uint64(pos))
+									}
+									txResp.Transaction.Transaction, txResp.Transaction.Meta, err = getTransactionAndMetaFromNode(txn, epochHandler.GetDataFrameByCid)
+									if err != nil {
+										errChan <- status.Errorf(codes.Internal, "Failed to get transaction: %v", err)
+										return
+									}
+									txResp.Slot = uint64(txn.Slot)
+
+									blocktimeIndex := epochHandler.GetBlocktimeIndex()
+									if blocktimeIndex != nil {
+										blocktime, err := blocktimeIndex.Get(uint64(txn.Slot))
+										if err != nil {
+											errChan <- status.Errorf(codes.Internal, "Failed to get blocktime: %v", err)
+											return
+										}
+										txResp.BlockTime = int64(blocktime)
+									} else {
+										errChan <- status.Errorf(codes.Internal, "Failed to get blocktime: blocktime index is nil")
+										return
+									}
 								}
-								txResp.BlockTime = int64(blocktime)
-							} else {
-								return status.Errorf(codes.Internal, "Failed to get blocktime: blocktime index is nil")
-							}
-						}
 
-						if err := ser.Send(txResp); err != nil {
-							return err
+								buffer.add(txResp.Slot, *txResp.Index, txResp)
+							}
 						}
+						buffer.markSlotComplete(slot)
```

Comment:
we don't know that we're done iwth the slot yet since we are so far just buffering for a single account